model/riscv_vmem_utils.sail

```diff
@@ -0,0 +1,225 @@
+/*=======================================================================================*/
+/*  This Sail RISC-V architecture model, comprising all files and                        */
+/*  directories except where otherwise noted is subject the BSD                          */
+/*  two-clause license in the LICENSE file.                                              */
+/*                                                                                       */
+/*  SPDX-License-Identifier: BSD-2-Clause                                                */
+/*=======================================================================================*/
+
+/* This file implements utility functions for accessing memory that
+ * can be used by instruction definitions.
+ */
+
+/* This is a external option that controls the order in which the model
+ * performs misaligned accesses.
+ */
+val sys_misaligned_order_decreasing = pure "sys_misaligned_order_decreasing" : unit -> bool
+
+/* This is an external option that, when true, causes all misaligned accesses
+ * to be split into single byte operations.
+ */
+val sys_misaligned_to_byte = pure "sys_misaligned_to_byte" : unit -> bool
+
+/* This is an external option that returns an integer N, such that
+ * when N is greater than zero, misaligned accesses to physical memory
+ * (as atomic events) are allowed provided the access occurs within a N
+ * byte region. N must be a power of two.
+ *
+ * This option cannot be unlimited (i.e. allowing all misaligned accesses
+ * to be atomic) as this would be incorrect above the maximum supported
+ * page size.
+ */
+val sys_misaligned_allowed_within = pure "sys_misaligned_allowed_within" : unit -> bits(64)
+
+/* The Zama16b extensions implies that misaligned loads, stores, and
+ * AMOs to main memory regions that do not cross a naturally aligned
+ * 16-byte boundary are atomic.
+ */
+enum clause extension = Ext_Zama16b
+
+val sys_enable_zama16b = pure "sys_enable_zama16b" : unit -> bool
+
+function clause extensionEnabled(Ext_Zama16b) = sys_enable_zama16b()
+
+/* Check if an 'n byte access for an address is within an aligned 'r byte region */
+val access_within : forall 'width 'n 'r, 1 <= 'n <= 'r. (bits('width), int('n), int('r)) -> bool
+
+function access_within(addr, bytes, region_bytes) = {
+  let low = addr & ~(get_slice_int(length(addr), bytes - 1, 0));
```

Comment:
Shouldn't this be
```suggestion
  let low = addr & ~(get_slice_int(length(addr), region_bytes - 1, 0));
```
The two property tests are testing `bytes == region_bytes` so they can't catch this issue.